warp_drive/training/trainer.py

```diff
@@ -249,6 +219,88 @@ def generate_rollout(self, batch_index):
         torch.cuda.synchronize()
         self.perf_stats.env_step_time += start_event.elapsed_time(end_event) / 1000
 
+    def evaluate_policies(self, batch_index):
+        probabilities = {}
+        for policy in self.policies:
+            probabilities[policy], _ = self.models[policy](
+                batch_index=batch_index, batch_size=self.training_batch_size_per_env
+            )
+            # Combine probabilities if there are multiple policies,
+            # yet they share the same placeholders
+            if (
+                len(self.policies) > 1
+                and not self.create_separate_placeholders_for_each_policy
+            ):
+                num_agents = self.cuda_envs.env.num_agents
+                agent_to_id_mapping = self.policy_tag_to_agent_id_map[policy]
+                combined_probabilities = [
+                    None for _ in range(len(probabilities[policy]))
+                ]
+                for action_idx, probs in enumerate(probabilities[policy]):
+                    num_envs = probs.shape[0]
+                    action_dim = probs.shape[-1]
+                    combined_probabilities[action_idx] = torch.zeros(
+                        (num_envs, num_agents, action_dim)
+                    ).cuda()
+                    combined_probabilities[action_idx][
+                        :, agent_to_id_mapping
+                    ] = probabilities[policy][action_idx]
+                probabilities[policy] = combined_probabilities
+        return probabilities
+
+    def sample_actions(self, probabilities, batch_index):
+        """
+        Sample action probabilities (and push the sampled actions to the device).
+        """
+        if self.create_separate_placeholders_for_each_policy:
+            for policy in self.policies:
+                suffix = f"_{policy}"
+                self.sample_actions_helper(
+                    probabilities[policy], batch_index, suffix=suffix
+                )
+        else:
+            for policy in self.policies:
+                suffix = ""
+            self.sample_actions_helper(
+                probabilities[policy], batch_index, suffix=suffix
+            )
+
+    def sample_actions_helper(self, probabilities, batch_index, suffix=""):
+        for action_idx, probs in enumerate(probabilities):
+            action_name = f"{_ACTIONS}_{action_idx}" + suffix
+            self.cuda_sample_controller.sample(
+                self.cuda_envs.cuda_data_manager, probs, action_name
+            )
+            actions = self.cuda_envs.cuda_data_manager.data_on_device_via_torch(
+                action_name
+            )
+
+            self.cuda_envs.cuda_data_manager.data_on_device_via_torch(
```

Comment:
If there is one single policy + one discrete action space, then there is no need to create a mapping buffer and copy it back. This will save time for example, shared policy for tag gridworld